clang/lib/AST/Interp/EvalEmitter.cpp

```diff
@@ -119,12 +121,26 @@ template <PrimType OpType> bool EvalEmitter::emitRet(const SourceInfo &Info) {
 template <> bool EvalEmitter::emitRet<PT_Ptr>(const SourceInfo &Info) {
   if (!isActive())
     return true;
-  EvalResult.setPointer(S.Stk.pop<Pointer>());
+
+  const Pointer &Ptr = S.Stk.pop<Pointer>();
+  // Implicitly convert lvalue to rvalue, if requested.
+  if (ConvertResultToRValue) {
+    if (std::optional<APValue> V = Ptr.toRValue(Ctx)) {
+      EvalResult.setValue(*V);
+    } else {
+      return false;
+    }
+  } else {
+    EvalResult.setPointer(Ptr);
+  }
+
   return true;
 }
 template <> bool EvalEmitter::emitRet<PT_FnPtr>(const SourceInfo &Info) {
   if (!isActive())
     return true;
+  // Function pointers are always lvalues to us and cannot be converted
+  // to rvalues, so don't do any conversion here.
```

Comment:
Whhaaat? Why are function pointers always lvalues?

Reply:
I guess that's not technically correct, but it's rather that the conversion for function pointers is never needed, regardless of whether they are lvalues or rvalues(?)

Reply:
I can't get any sample to hit the `evaluteAsRValue()` code path:
```c
int foo() { return 0; }
auto f = foo;
const auto ff = &foo;
```
even for C. I can add an `assert(!ConvertResultToRValue)` if that makes sense.

Reply:
Okay, I think an assert seems reasonable; that should tell us if our assumptions are off-base.

Reply:
Nothing in `test/Sema/` seems to trip that assertion up (but that's just C). Note that there's also this comment in `ExprConstant.cpp`:

```c
  // In C, function designators are not lvalues, but we evaluate them as if they
  // are.
```

Reply:
Well, there's a difference between a function designator and a function pointer... Given:
```
void func(int);

int main() {
  func(0);
}
```
`func` is a function designator, the `func(0)` decays the function designator `void(int)` into a function pointer `void (*)(int)` which is then used by the call expression and passed the integer literal `0`. In terms of our AST:
```
    `-CallExpr <line:4:3, col:9> 'void'
      |-ImplicitCastExpr <col:3> 'void (*)(int)' <FunctionToPointerDecay>
      | `-DeclRefExpr <col:3> 'void (int)' Function 0xc59b980 'func' 'void (int)'
      `-IntegerLiteral <col:8> 'int' 0
```
The `DeclRefExpr` is using the function designator, the `ImplicitCastExpr` is what turns that into a function pointer, which is then the first operand of the `CallExpr`, and the `IntegerLiteral` is the first formal argument to the `CallExpr`.

I think we evaluate function designators as-if they are lvalues because they're a `DeclRefExpr` in terms of the call expression and those are lvalues until there's a conversion on them.

Reply:
I've reworded the comment.